serving/docs/adapters.md

```diff
@@ -0,0 +1,180 @@
+    -H "Content-Type: application/json" \
+    -H "X-Amzn-SageMaker-Target-Model: base-1.tar.gz" \
+    -d '{"inputs": ["How is the weather"], "adapters": ["adapter_1"], "parameters": {"max_new_tokens": 25}}'
+```
+
+### Input Content (and query parameter)
+
+This passes the adapter through a query parameter.
+It is reflected in the Input content.
+This will not work in Amazon SageMaker.
+
+```
+curl -X POST http://127.0.0.1:8080/invocations?adapter=adapter_1 \
+    -H "Content-Type: application/json" \
+    -H "X-Amzn-SageMaker-Target-Model: base-1.tar.gz" \
+    -d '{"inputs": ["How is the weather"], "parameters": {"max_new_tokens": 25}}'
+```
+
+### SageMaker Custom Attributes
+
+This passes the adapter through the use of the SageMaker Custom Attributes header.
+It is reflected in the Input properties.
+This option will work in Amazon SageMaker.
+
+```
+curl -X POST http://127.0.0.1:8080/invocations \
+    -H "Content-Type: application/json" \
+    -H "X-Amzn-SageMaker-Target-Model: base-1.tar.gz" \
+    -H "X-Amzn-SageMaker-Custom-Attributes: adapter=adapter_1"
+    -d '{"inputs": ["How is the weather"], "parameters": {"max_new_tokens": 25}}'
+```
+
+## Models with Adapters
+
+There are two kinds of LoRA that can be put onto various engines.
+
+* **Merged LoRA** - This will apply the adapter by modifying the base model in place.
+  It has zero added latency during execution, but has a cost to apply or unapply the merge.
+  It works best for cases with only a few adapters.
+  It is best for single adapter batches, but doesn’t support multi-adapter batches
+* **Unmerged LoRA** - This will alter the model operators to factor in the adapters without changing the base model.
+  It has a higher inference latency for the additional adapter operations.
+  However, it does support multi-adapter batches.
+  It works best for use cases with large numbers of adapters.
+
+With our default handlers, we currently support unmerged LoRA for CausalLM through the huggingface handler.
+Support for other model types and handlers is coming soon.
+
+### Writing Adapter Models
+
+Right now, adapters are only supported through our Python engine.
+See instructions to get started with writing a [python handler](modes.md#python-mode).
+
+To add support for adapters, you must first add the register and unregister like below.
+These can then take the adapters and save the src, pre-download it, cache it in memory, or cache it on an accelerator device.
+
+```python
+def register_adapter(inputs: Input):
+  name = inputs.get_properties()["name"]
+  src = inputs.get_properties()["src"]
+  # Do adapter registration tasks
+  return Output().add("Successfully registered adapter")
+
+def unregister_adapter(inputs: Input):
+  name = inputs.get_properties()["name"]
+  # Do adapter unregistration tasks
+  return Output().add("Successfully unregistered adapter")
+
+
+def handle(inputs: Input):
+  ...
+```
+
+Within the handler, you must parse the adapter properties from the inputs.
```

Comment:
we should also add that `adapters` arg should be passed to `generate` function as a `kwarg`. For e.g
```
generate(input_ids=input_tokens.input_ids,
                attention_mask=input_tokens.attention_mask,
                adapters=[adapter_1, adapter_2],
               **kwargs
)

```

Reply:
Can you give more context about the generate function? Is this PySDK or something?